src/models/RuleServer.ts

```diff
@@ -0,0 +1,311 @@
+/*
+Copyright 2019-2021 The Matrix.org Foundation C.I.C.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+import BanList, { ChangeType, ListRuleChange, RULE_ROOM, RULE_SERVER, RULE_USER } from "./BanList"
+import * as crypto from "crypto";
+import { LogService } from "matrix-bot-sdk";
+import { ListRule } from "./ListRule";
+
+export const USER_MAY_INVITE = 'user_may_invite';
+export const CHECK_EVENT_FOR_SPAM = 'check_event_for_spam';
+
+/**
+ * Rules in the RuleServer format that have been produced from a single event.
+ */
+class EventRules {
+    constructor (
+        readonly eventId: string,
+        readonly roomId: string,
+        readonly ruleServerRules: RuleServerRule[],
+        // The token associated with when the event rules were created.
+        readonly token: number
+    ) {
+    }
+}
+
+/**
+ * A description of a property that should be checked as part of a RuleServerRule.
+ */
+interface Checks {
+    property: string;
+}
+
+/**
+ * A Rule served by the rule server.
+ */
+interface RuleServerRule {
+    // A unique identifier for this rule.
+    readonly id: string
+    // Human-readable description, for audit purposes
+    readonly description?: string
+    // A description of a property that should be checked.
+    readonly checks: Checks
+}
+
+/**
+ * The RuleServer is an experimental server that is used to propogate the rules of the watched policy rooms (BanLists) to
+ * homeservers (or e.g. synapse modules).
+ * This is done using an experimental format that is heavily based on the "Spam Checker Callbacks" made available to
+ * synapse modules https://matrix-org.github.io/synapse/latest/modules/spam_checker_callbacks.html.
+ *
+ */
+export default class RuleServer {
+    // Each token is an index for a row of this two dimensional array. Each row represents the rules that were added during the lifetime of that token.
+    private ruleStartsByToken: EventRules[][] = [[]];
+    // Each row, indexed by a token, represents the rules that were stopped during the lifetime of that token.
+    private ruleStopsByToken: string[][] = [[]];
+
+    // We use this to quickly lookup if we have stored a policy without scanning rulesByToken.
+    // First key is the room id and the second is the event id.
+    private rulesByEvent: Map<string, Map<string, EventRules>> = new Map();
```

Comment:
For `get` this doesn't make sense because you call this with an event not an EventRules instance like with `addEventRules` and `stopEventRules`. `addEventRules` and `stopEventRules` also operate on other members than `rulesByEvent` so it is also problematic to refactor in this way.